Fix compilation on GCC with -Wold-style-cast, -Wsign-conversion and -Wzero-as-null-pointer-constant

ArekPiekarz opened the pull request:
This has been tested locally on Ubuntu 17.10 with GCC 7.2 (all 64-bit).

bombela commented:
Please rebase and double check that it's still warning free. I merged a big PR. Thank you for contributing.

ArekPiekarz commented:
Thanks, I rebased the commits.

ArekPiekarz commented:
I noticed that on Clang 5 when you use C++98 mode and enable `-Weverything` then a warning `-Wnon-literal-null-conversion` appears in lines where I replaced zeroes with details::null_ptr. What do you want me to do with it? I see these options:

1) Ignore it and merge the pull request.
2) Don't merge until it is fixed, maybe using ifdefs?
3) Split the pull request into fixes for each warning, then merge the ones for `-Wold-style-cast` and `-Wsign-conversion`, wait for fix for `-Wnon-literal-null-conversion`.
4) Something else?

Here are the files I used for testing from C++98 to C++17 for reference:

CMakeLists.txt:
```
cmake_minimum_required(VERSION 3.0)
project(backward-cpp-fork-example LANGUAGES CXX)

set(CMAKE_CXX_STANDARD 17)  # change to test other standards: 98, 11, 14, 17
set(CMAKE_CXX_STANDARD_REQUIRED ON)
set(CMAKE_CXX_EXTENSIONS OFF)

if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
    
    add_compile_options(
        -Werror
        -Wpedantic
        -pedantic-errors
        -Wall
        -Wextra
        -Wchkp
        -Wdouble-promotion
        -Wformat=2
        -Wformat-overflow=2
        -Wformat-signedness
        -Wformat-truncation=2
        -Wnull-dereference
        -Wimplicit-fallthrough=5
        -Wmissing-include-dirs
        -Wshift-overflow=2
        -Wswitch-enum
        -Wsync-nand
        -Wunused-parameter
        -Wunused-const-variable=1
        -Wunknown-pragmas
        -Wstrict-overflow=5
        -Wstringop-overflow=4
        -Wsuggest-attribute=pure
        -Wsuggest-attribute=const
        -Wsuggest-attribute=noreturn
        -Wsuggest-attribute=format
        -Wmissing-format-attribute
        -Wsuggest-final-types
        -Wsuggest-final-methods
        -Walloc-zero
        -Walloca
        -Warray-bounds=2
        -Wduplicated-branches
        -Wduplicated-cond
        -Wtrampolines
        -Wfloat-equal
        -Wshadow
        -Wunsafe-loop-optimizations
        -Waligned-new=all
        -Wplacement-new=2
        -Wundef
        -Wunused-macros
        -Wcast-qual
        -Wcast-align
        -Wconditionally-supported
        -Wconversion
        -Wdate-time
        -Wuseless-cast
        -Wsign-conversion
        -Wsized-deallocation
        -Wlogical-op
        -Wmissing-declarations
        -Wnormalized=id
        -Wpacked
        -Wredundant-decls
        -Wrestrict
        -Winvalid-pch
        -Wvector-operation-performance
        -Wdisabled-optimization
        -Wstack-protector
        -Wno-unknown-pragmas
        # Additional warnings for C++ dialect
        -Wctor-dtor-privacy
        -Wnoexcept
        -Wnon-virtual-dtor
        -Wstrict-null-sentinel
        -Wold-style-cast
        -Woverloaded-virtual
        -Wsign-promo
    )
    
    if (NOT CMAKE_CXX_STANDARD EQUAL 98)
        add_compile_options(
            -Wsuggest-override
            -Wzero-as-null-pointer-constant
        )
    endif()
    
elseif (CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
    
    add_compile_options(
        -Weverything
        -Werror
        -Wno-padded
        -Wno-weak-vtables
    )

    if (CMAKE_CXX_STANDARD EQUAL 98)
        add_compile_options(
            -Wno-non-literal-null-conversion
        )
    else()
        add_compile_options(
            -Wno-c++98-compat
            -Wno-c++98-compat-pedantic
        )
    endif()
    
endif()

list(APPEND CMAKE_PREFIX_PATH external/backward-cpp)
find_package(Backward REQUIRED)
add_executable(${PROJECT_NAME} "main.cpp")
target_link_libraries(${PROJECT_NAME} PUBLIC Backward::Backward)
```

main.cpp:
```
#include <csignal>
// below is required when using libbfd library because of its macros
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wzero-as-null-pointer-constant"
#pragma clang diagnostic ignored "-Wdisabled-macro-expansion"
#include "external/backward-cpp/backward.hpp"
#pragma clang diagnostic pop

int main()
{
    backward::SignalHandling signalHandling;
    raise(SIGSEGV);
    return 0;
}
```

ArekPiekarz commented:
I fixed the problem with `-Wnon-literal-null-conversion` by adding a macro `#define nullptr NULL` only for C++98 mode and replacing usages of `details::null_ptr` with `nullptr`.

ArekPiekarz commented:
I also now silenced on Clang:
- `-Wzero-as-null-pointer-constant` warning in macro from libbfd
- `-Wdisabled-macro-expansion` warning in macro from signal.h

bombela commented:
I've used the online editor to merge master back.
